src/mumble/widgets/TrayIcon.cpp

```diff
@@ -0,0 +1,208 @@
+
+	// messageClicked is buggy in Qt on some platforms and we can not do anything about this (QTBUG-87329)
+	QObject::connect(this, &QSystemTrayIcon::messageClicked, this, &TrayIcon::on_showAction_triggered);
+
+	m_showAction = new QAction(tr("Show"), Global::get().mw);
+	QObject::connect(m_showAction, &QAction::triggered, this, &TrayIcon::on_showAction_triggered);
+
+	m_hideAction = new QAction(tr("Hide"), Global::get().mw);
+	QObject::connect(m_hideAction, &QAction::triggered, this, &TrayIcon::on_hideAction_triggered);
+
+	QObject::connect(Global::get().mw->qaTalkingUIToggle, &QAction::triggered, this, &TrayIcon::updateContextMenu);
+
+	m_contextMenu = new QMenu(Global::get().mw);
+	QObject::connect(m_contextMenu, &QMenu::aboutToShow, this, &TrayIcon::updateContextMenu);
+
+	// Some window managers hate it when a tray icon sets an empty context menu...
+	updateContextMenu();
+
+	setContextMenu(m_contextMenu);
+
+	show();
+}
+
+void TrayIcon::updateIcon() {
+	QIcon *newIcon = nullptr;
+
+	ClientUser *p = ClientUser::get(Global::get().uiSession);
+
+	if (Global::get().s.bDeaf) {
+		newIcon = &Global::get().mw->qiIconDeafSelf;
+	} else if (p && p->bDeaf) {
+		newIcon = &Global::get().mw->qiIconDeafServer;
+	} else if (Global::get().s.bMute) {
+		newIcon = &Global::get().mw->qiIconMuteSelf;
+	} else if (p && p->bMute) {
+		newIcon = &Global::get().mw->qiIconMuteServer;
+	} else if (p && p->bSuppress) {
+		newIcon = &Global::get().mw->qiIconMuteSuppressed;
+	} else if (Global::get().s.bStateInTray && Global::get().bPushToMute) {
+		newIcon = &Global::get().mw->qiIconMutePushToMute;
+	} else if (p && Global::get().s.bStateInTray) {
+		switch (p->tsState) {
+			case Settings::Talking:
+			case Settings::MutedTalking:
+				newIcon = &Global::get().mw->qiTalkingOn;
+				break;
+			case Settings::Whispering:
+				newIcon = &Global::get().mw->qiTalkingWhisper;
+				break;
+			case Settings::Shouting:
+				newIcon = &Global::get().mw->qiTalkingShout;
+				break;
+			case Settings::Passive:
+			default:
+				newIcon = &Global::get().mw->qiTalkingOff;
+				break;
+		}
+	} else {
+		newIcon = &Global::get().mw->qiIcon;
+	}
+
+	m_iconLock.lockForRead();
+	bool isStatusIcon = newIcon == m_statusIcon;
+	m_iconLock.unlock();
+
+	if (!isStatusIcon) {
+		m_iconLock.lockForWrite();
+		m_statusIcon = newIcon;
+		m_iconLock.unlock();
+
+		applyIcon(m_statusIcon);
+	}
+}
+
+void TrayIcon::on_icon_clicked(QSystemTrayIcon::ActivationReason reason) {
+	switch (reason) {
+		case QSystemTrayIcon::Trigger:
+#ifndef Q_OS_MAC
+			// macOS is special as it both shows the context menu AND triggers the action.
+			// We only want at most one of those and since we can not prevent showing
+			// the menu, we skip the action.
+			toggleShowHide();
+#endif
+			break;
+		case QSystemTrayIcon::Unknown:
+		case QSystemTrayIcon::Context:
+		case QSystemTrayIcon::DoubleClick:
+		case QSystemTrayIcon::MiddleClick:
+		default:
```

Comment:
Why have all these cases explicitly listed when they all fall-through to the `default` clause. If the enumerated cases exhaust the possible values for `reason`, we should remove the `default` clause in order to get compiler warnings about uncovered cases as soon as Qt introduces new values into the enum.